File: app/src/main/java/com/balicodes/quicksms/RecipientListAdapter.java
package com.balicodes.quicksms;

import android.content.Context;
import android.content.Intent;
import android.provider.ContactsContract;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by eka on 6/18/16.
 */
public class RecipientListAdapter extends BaseAdapter {
    private List<String[]> items;
    private SMSFormFragment fragment;

    public RecipientListAdapter(SMSFormFragment fragment, List<String[]> items) {
        this.fragment = fragment;
        this.items = items;
    }

    public void addItem(String[] item) {
        items.add(item);
        notifyDataSetChanged();
    }

    public void removeItem(int position) {
        items.remove(position);
        notifyDataSetChanged();
    }

    public void pickNumber(int position) {
        Intent pickContactIntent = new Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI);
        pickContactIntent.setType(ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE);
        fragment.setRecipientPickIndex(position);
        fragment.startActivityForResult(pickContactIntent, Config.PICK_CONTACT_REQUEST);
    }

    public void addContactToItem(int position, String[] recipient) {
        items.set(position, recipient);
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public Object getItem(int position) {
        return items.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) fragment.getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.receiver_list_item, parent, false);
        }
        // Remove item button
        ImageButton removeNumberBtn = (ImageButton) convertView.findViewById(R.id.removeNumberBtn);
        removeNumberBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (getCount() == 1) {
                    Toast.makeText(fragment.getActivity(), R.string.number_error, Toast.LENGTH_LONG).show();
                    return;
                }
                removeItem(position);
            }
        });

        // pick a number
        ImageButton pickNumberBtn = (ImageButton) convertView.findViewById(R.id.pickNumberBtn);
        pickNumberBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                pickNumber(position);
            }
        });

        final EditText recName = (EditText) convertView.findViewById(R.id.recName);
        final EditText recNumber = (EditText) convertView.findViewById(R.id.recNumber);

        try {
            String[] receiver = items.get(position);
            recName.setText(receiver[0]);
            recNumber.setText(receiver[1]);
        } catch (ArrayIndexOutOfBoundsException e) {
            e.printStackTrace();
        }

        return convertView;
    }
}

File: app/src/main/java/com/balicodes/quicksms/RecipientListAdapter.kt
/*
 * Copyright (C) 2016-2018 Eka Putra
 *
 * Quick SMS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.balicodes.quicksms

import android.content.Context
import android.content.Intent
import android.provider.ContactsContract
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Toast

internal class RecipientListAdapter(private val fragment: SMSFormFragment, private val items: MutableList<Array<String>>) : BaseAdapter() {

    fun addItem(item: Array<String>) {
        items.add(item)
        notifyDataSetChanged()
    }

    private fun removeItem(position: Int) {
        items.removeAt(position)
        notifyDataSetChanged()
    }

    private fun pickNumber(position: Int) {
        val pickContactIntent = Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI)
        pickContactIntent.type = ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE
        fragment.setRecipientPickIndex(position)
        fragment.startActivityForResult(pickContactIntent, Config.PICK_CONTACT_REQUEST)
    }

    fun addContactToItem(position: Int, recipient: Array<String>) {
        items[position] = recipient
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return items.size
    }

    override fun getItem(position: Int): Any {
        return items[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        if (convertView == null) {
            val inflater = fragment.activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            convertView = inflater.inflate(R.layout.receiver_list_item, parent, false)
        }
        // Remove item button
        val removeNumberBtn = convertView!!.findViewById<ImageButton>(R.id.removeNumberBtn)
        removeNumberBtn.setOnClickListener(View.OnClickListener {
            if (count == 1) {
                Toast.makeText(fragment.activity, R.string.number_error, Toast.LENGTH_LONG).show()
                return@OnClickListener
            }
            removeItem(position)
        })

        // pick a number
        val pickNumberBtn = convertView.findViewById<ImageButton>(R.id.pickNumberBtn)
        pickNumberBtn.setOnClickListener { pickNumber(position) }

        val recName = convertView.findViewById<EditText>(R.id.recName)
        val recNumber = convertView.findViewById<EditText>(R.id.recNumber)

        try {
            val receiver = items[position]
            recName.setText(receiver[0])
            recNumber.setText(receiver[1])
        } catch (e: ArrayIndexOutOfBoundsException) {
            e.printStackTrace()
        }

        return convertView
    }
}
